server/storage/backend/backend.go

```diff
@@ -477,29 +477,38 @@ func (b *backend) defrag() error {
 	b.readTx.Lock()
 	defer b.readTx.Unlock()
 
-	b.batchTx.unsafeCommit(true)
-
-	b.batchTx.tx = nil
-
 	// Create a temporary file to ensure we start with a clean slate.
 	// Snapshotter.cleanupSnapdir cleans up any of these that are found during startup.
 	dir := filepath.Dir(b.db.Path())
 	temp, err := os.CreateTemp(dir, "db.tmp.*")
 	if err != nil {
 		return err
 	}
+	defer temp.Close()
+
+	// Commit/stop and then reset current transactions (including the readTx)
+	b.batchTx.unsafeCommit(true)
+	b.batchTx.tx = nil
```

Comment:
We can even move these two lines to right before `defragdb`. i.e. if we fail to open the temp bbolt db, we don't need to reopen the existing working bbolt db.

https://github.com/etcd-io/etcd/blob/bd8896364a1d9adb554b538232004e5b2d6c850a/server/storage/backend/backend.go#L519

Reply:
Of course, yes 🙂, will move it.

Reply:
Now in case the `defragdb` fails and returns an error we could potentially endup in the same state as described in the issue (`b.batchTx.tx == nil` and nil ptr panic). Should this error be handled or is this something were etcd should stop e.g. panic?

Reply:
Please see https://github.com/etcd-io/etcd/pull/18822#issuecomment-2459209181

Reply:
I added a second commit for that, please let me know if that is ok.